source-map.bs

```diff
@@ -256,14 +366,172 @@ with four fields represent mapped code where a corresponding name does not exist
+    1. If |position| points to the end of |segment|, throw an error.
+    1. Set |currentByte| to the [=byte=] whose the [=byte/value=] is the number corresponding to
+        |segment|'s |position|th [=code unit=], according to the [[base64]] encoding.
+    1. Let |chunk| be |currentByte| & 0x1F, as a number.
+    1. Add |chunk| * |nextShift| to |value|.
+    1. If |value| is greater than or equal to 2<sup>31</sup>, throw an error.
+    1. Multiply |nextShift| by 32.
+1. Advance |position| by 1.
+1. If |value| is 0 and |sign| is -1, return -2147483648.
+
+    NOTE: -2147483648 is the smallest 32-bit signed integer.
+1. Return |value| * |sign|.
+
+NOTE: In addition to returning the decoded value, this algorithm updates the [=position variable=]
+in the calling algorithm.
 
 Resolving Sources {#resolving-sources}
 --------------------------------------
 
-If the sources are not absolute URLs after prepending the [=sourceRoot=], the sources are
+If the sources are not absolute URLs after prepending the [=json/sourceRoot=], the sources are
 resolved relative to the SourceMap (like resolving the script `src` attribute in an HTML document).
 
+To <dfn>decode source map sources</dfn> given a [=/URL=] |baseURL|,
+a [=string=] or null <dfn for="decode source map sources">|sourceRoot|</dfn>,
+a [=list=] of either [=strings=] or nulls <dfn for="decode source map sources">|sources|</dfn>,
+a [=list=] of either [=strings=] or nulls <dfn for="decode source map sources">|sourcesContent|</dfn>,
+and a [=list=] of numbers <dfn for="decode source map sources">|ignoredSources|</dfn>,
+run the following steps:
+1. Let |decodedSources| be a new empty [=list=].
+1. Let |sourceURLPrefix| be "".
+1. If |sourceRoot| is not null, then:
+    1. If |sourceRoot| contains the code point U+002F (/), then:
+        1. Let |index| be the index of the last occurrence of U+002F (/) in |sourceRoot|.
+        1. Set |sourceURLPrefix| to the [=code unit substring|substring=] of |sourceRoot| from 0 to
+            |index| + 1.
+    1. Else, set |sourceURLPrefix| to the concatenation of |sourceRoot| and "/".
+1. [=For each=] |source| of |sources| with index |index|:
+    1. Let |decodedSource| be a new [=decoded source=] whose [=decoded source/URL=] is null,
+        [=decoded source/content=] is null, and [=decoded source/ignored=] is false.
+    1. If |source| is not null:
+        1. Set |source| to the [=string/concatenate|concatenation=] of |sourceURLPrefix| and |source|.
+        1. Let |sourceURL| be the result of [=URL parser|URL parsing=] |source| with |baseURL|.
+        1. If |sourceURL| is failure, [=optionally report an error=].
+        1. Else, set |decodedSource|'s [=decoded source/URL=] to |sourceURL|.
+    1. If |index| is in |ignoredSources|, set |decodedSource|'s [=decoded source/ignored=] to true.
+    1. If |sourcesContent|'s [=list/size=] is greater than or equal to |index|, set
+        |decodedSource|'s [=decoded source/content=] to |sourcesContent|[|index|].
```

Comment:
The handling of sourcesContent here seems like a natural way to spec it, but when you have duplicate entries it currently conflicts with what implementations actually do. (see [this issue](https://github.com/tc39/source-map/issues/85) for more details) In particular, it seems all the major browsers treat sources entries as being set-like so that duplicate source URLs point to the same sourcesContent (either the first or last one for that particular URL).